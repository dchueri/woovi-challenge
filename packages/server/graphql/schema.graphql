"""Root query"""
type Query {
  movies(
    """Returns the items in the list that come after the specified cursor."""
    after: String

    """Returns the first n items from the list."""
    first: Int

    """Returns the items in the list that come before the specified cursor."""
    before: String

    """Returns the last n items from the list."""
    last: Int
  ): MovieConnection!
  user(recovery: String!): User
}

"""A connection to a list of items."""
type MovieConnection {
  """Information to aid in pagination."""
  pageInfo: PageInfo!

  """A list of edges."""
  edges: [MovieEdge]
}

"""Information about pagination in a connection."""
type PageInfo {
  """When paginating forwards, are there more items?"""
  hasNextPage: Boolean!

  """When paginating backwards, are there more items?"""
  hasPreviousPage: Boolean!

  """When paginating backwards, the cursor to continue."""
  startCursor: String

  """When paginating forwards, the cursor to continue."""
  endCursor: String
}

"""An edge in a connection."""
type MovieEdge {
  """The item at the end of the edge"""
  node: Movie

  """A cursor for use in pagination"""
  cursor: String!
}

"""Movie Type"""
type Movie {
  id: String
  title: String
  genre: String
  image: String
  description: String
  average: Float
}

type User implements Node {
  """The ID of an object"""
  id: ID!
  name: String!
  email: String!
  recovery: String!
  helperSeen: Boolean
}

"""An object with an ID"""
interface Node {
  """The id of the object."""
  id: ID!
}

type Mutation {
  CreateMovie(input: CreateMovieInput!): CreateMoviePayload
  DeleteMovie(input: DeleteMovieInput!): DeleteMoviePayload
  UpdateMovie(input: UpdateMovieInput!): UpdateMoviePayload
  LoginMutation(input: UserLoginInput!): UserLoginPayload
  RegisterUserMutation(input: UserRegisterInput!): UserRegisterPayload
  UpdateUserMutation(input: UpdateUserInput!): UpdateUserPayload
  CommentCreateMutation(input: CommentCreateInput!): CommentCreatePayload
}

type CreateMoviePayload {
  movieEdge: MovieEdge
  error: String
  clientMutationId: String
}

input CreateMovieInput {
  title: String!
  genre: String!
  image: String!
  description: String!
  average: Float!
  clientMutationId: String
}

type DeleteMoviePayload {
  error: String
  deletedId: String
  clientMutationId: String
}

input DeleteMovieInput {
  id: String!
  clientMutationId: String
}

type UpdateMoviePayload {
  movieEdge: MovieEdge
  error: String
  clientMutationId: String
}

input UpdateMovieInput {
  id: String!
  title: String
  genre: String
  image: String
  description: String
  average: Float
  clientMutationId: String
}

type UserLoginPayload {
  token: String
  me: User
  error: String
  clientMutationId: String
}

input UserLoginInput {
  email: String!
  password: String!
  clientMutationId: String
}

type UserRegisterPayload {
  token: String
  me: User
  success: String
  error: String
  clientMutationId: String
}

input UserRegisterInput {
  name: String!
  email: String!
  password: String!
  clientMutationId: String
}

type UpdateUserPayload {
  success: String
  error: String
  clientMutationId: String
}

input UpdateUserInput {
  email: String!
  name: String
  password: String
  recovery: String
  helperSeen: Boolean
  clientMutationId: String
}

type CommentCreatePayload {
  commentEdge: CommentEdge
  movie: Movie
  error: String
  clientMutationId: String
}

"""An edge in a connection."""
type CommentEdge {
  """The item at the end of the edge."""
  node: Comment

  """A cursor for use in pagination."""
  cursor: String!
}

"""Comment data"""
type Comment implements Node {
  """The ID of an object"""
  id: ID!
  text: String
  user: User
  movie: Movie
  createdAt: String
  updatedAt: String
}

input CommentCreateInput {
  movie: ID!
  text: String!
  clientMutationId: String
}