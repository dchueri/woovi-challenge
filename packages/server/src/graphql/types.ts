import mongoose from "mongoose";
import { IUser } from "../types";

import { DataLoaders } from "../modules/loader/loaderRegister";

declare type ObjectId = mongoose.Schema.Types.ObjectId;

export type GraphQLContext = {
  user?: IUser | null;
  dataloaders: DataLoaders;
};

export type LoaderFn = (
  ctx: GraphQLContext,
  id: string | ObjectId | object
) => any;
