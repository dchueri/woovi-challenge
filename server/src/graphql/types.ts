import mongoose from "mongoose";
import { IUser } from "../types/types";

import { DataLoaders } from "../graphql/loaderRegister";

declare type ObjectId = mongoose.Schema.Types.ObjectId;

export type GraphQLContext = {
  user?: IUser;
  dataloaders: DataLoaders;
};

export type LoaderFn = (
  ctx: GraphQLContext,
  id: string | ObjectId | object
) => any;
