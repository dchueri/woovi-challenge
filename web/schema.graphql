type Movie {
  movieEdge: MovieEdge
  error: String
  id: String!
  title: String!
  genre: String!
  image: String
}

type Node {
  id: String!
  title: String!
  genre: String!
  image: String
}

type MovieEdge {
  node: Node
  cursor: String
}

type IMovieEdge {
  node: Node
  cursor: String
}

input CreateMovieInput {
  title: String!
  genre: String!
  image: String
}

input DeleteMovieInput {
  id: String!
}

input LoginInput {
  email: String!
  password: String!
}

input RegisterUserInput {
  name: String!
  email: String!
  password: String!
}

type Mutation {
  CreateMovie(input: CreateMovieInput): Movie
  DeleteMovie(input: DeleteMovieInput): DeleteMovie
  LoginMutation(input: LoginInput): LoginOrRegister
  RegisterUserMutation(input: RegisterUserInput): LoginOrRegister
  UpdateUserMutation(input: UpdateUserInput): SimpleReturn
}

type Query {
  movies(first: Int): Movies
  user(recovery: String!): User
}

type SimpleReturn {
  success: String
  error: String
}

input UpdateUserInput {
  email: String!
  name: String
  password: String
  recovery: String
}

input MoviesQueryInput {
  after: String
  before: String
  first: Int
  last: Int
}

type LoginOrRegister {
  token: String
  me: User
  error: String
}

type User {
  id: String
  name: String
  email: String
}

type DeleteMovie {
  deletedId: ID
  error: String
}

type Movies {
  edges: [IMovieEdge]
  pageInfo: PageInfo
}

type PageInfo {
  hasNextPage: String
  hasPreviousPage: String
  startCursor: String
  endCursor: String
}
