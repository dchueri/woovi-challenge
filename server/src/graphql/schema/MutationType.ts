import { GraphQLObjectType } from "graphql";
import MovieMutations from "../../modules/movie/mutations";
import UserMutations from "../../modules/user/mutations";

const MutationType = new GraphQLObjectType({
  name: "Mutation",
  fields: () => ({
    ...MovieMutations,
    ...UserMutations,
  }),
});

export default MutationType;
