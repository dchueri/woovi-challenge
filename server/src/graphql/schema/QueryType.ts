import { GraphQLNonNull, GraphQLObjectType } from "graphql";
import { connectionArgs, connectionFromArray } from "graphql-relay";
import * as MovieLoader from "../../graphql/movie/MovieLoader";
import { MovieConnection } from "../../graphql/movie/MovieType";

const QueryType = new GraphQLObjectType({
  name: "Query",
  description: "Root query",
  fields: () => ({
    movies: {
      type: new GraphQLNonNull(MovieConnection),
      args: connectionArgs,
      resolve: async (_, args, context) => {
        const data = await MovieLoader.getAll();
        return connectionFromArray(data, args);
      },
    },
  }),
});

export default QueryType;
