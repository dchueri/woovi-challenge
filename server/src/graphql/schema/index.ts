import { GraphQLSchema } from "graphql";
import MutationType from "../../graphql/schema/MutationType";
import QueryType from "../../graphql/schema/QueryType";

export const schema = new GraphQLSchema({
  query: QueryType,
  mutation: MutationType,
});
