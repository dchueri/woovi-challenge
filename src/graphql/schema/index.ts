import { GraphQLSchema } from "graphql";
import MutationType from "./MutationType";
import QueryType from "./QueryType";

export const schema = new GraphQLSchema({
  query: QueryType,
  mutation: MutationType,
});
