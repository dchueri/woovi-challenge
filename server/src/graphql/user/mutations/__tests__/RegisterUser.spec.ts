/* const EasyGraphQLTester = require("easygraphql-tester");
import { graphql } from "graphql";
import { MongoClient } from "mongodb";
import { schema } from "../../../schema";
const fs = require("fs");
const path = require("path");

//const userSchema = fs.readFileSync(path.join(__dirname, 'schema', 'user.gql'), 'utf8')

const tester = new EasyGraphQLTester(schema);

describe("insert", () => {
  let connection;
  let db;

  beforeAll(async () => {
    connection = await await MongoClient.connect(global.__MONGO_URI__);
    db = await connection.db();
  });

  afterAll(async () => {
    await connection.close();
  });

  it("should create a user", async () => {
    const users = db.collection("JestUsers");

    const mutation = `
      mutation RegisterUserMutation($name: String!, $email: String!, $password: String!) {
        RegisterUserMutation(input: {name: $name, email: $email, password: $password}) {
          token
          me {
            name
            email
          }
        }
      }
    `;

    const mockUser = {
      name: "Test",
      email: "test@test.com",
      password: "123456",
    };

    const result = await graphql({
      schema: schema,
      source: mutation,
      variableValues: { ...mockUser },
    });
    console.log(result);

    expect(result.data.me).toBeDefined();
  });
}); */

/* import { graphql } from "graphql";

import {
  connectWithMongoose,
  disconnectWithMongoose
} from "../../../../../test";
import { schema } from "../../../schema";

beforeAll(connectWithMongoose);
afterAll(disconnectWithMongoose);

it("should create a new user", async () => {
  const mutation = `
      mutation RegisterUserMutation($name: String!, $email: String!, $password: String!) {
        RegisterUserMutation(input: {name: $name, email: $email, password: $password}) {
          token
          me {
            id
            name
            email
          }
        }
      }
    `;

  const rootValue = {};

  const variableValues = {
    name: "test",
    email: "test@test.com",
    password: "123456",
  };

  const result = await graphql({
    schema,
    source: mutation,
    rootValue,
    variableValues,
  });

  console.log(result)
  expect(result.errors).toBeUndefined();

  const { me, token } = result?.data?.registerUserMutation as any;

  expect(token).toBeDefined();
  expect(me.id).toBeDefined();
}); */

import { graphql } from "graphql";
import {
  clearDbAndRestartCounters,
  connectMongoose,
  disconnectMongoose,
  sanitizeTestObject
} from "../../../../../test";
import { schema } from "../../../schema";
import { createUser } from "../../fixture/createUser";

beforeAll(connectMongoose);

beforeEach(clearDbAndRestartCounters);

afterAll(disconnectMongoose);

it("should register a user", async () => {
  const mutation = `
  mutation RegisterUserMutation($input: UserRegisterInput!) {
    RegisterUserMutation(input: $input) {
      token
      me {
        id
        name
        email
      }
      error
    }
  }
  `;

  const variableValues = {
    input: { name: "test", email: "test@test.com", password: "123456" },
  };
  const rootValue = {};

  const result = await graphql({
    schema,
    source: mutation,
    rootValue,
    variableValues,
  });

  expect(result.errors).toBeUndefined();
  expect(result.data.RegisterUserMutation.error).toBeNull();
  expect(result.data.RegisterUserMutation.me.name).toEqual(
    variableValues.input.name
  );
  expect(sanitizeTestObject(result.data)).toMatchSnapshot();
});

it("should not register a user when email already exists", async () => {
  const mutation = `
  mutation RegisterUserMutation($input: UserRegisterInput!) {
    RegisterUserMutation(input: $input) {
      token
      me {
        id
        name
        email
      }
      error
    }
  }
  `;

  const user = await createUser();

  const variableValues = {
    input: { name: "test", email: user.email, password: "123456" },
  };
  const rootValue = {};

  const result = await graphql({
    schema,
    source: mutation,
    rootValue,
    variableValues,
  });

  expect(result.errors).toBeUndefined();
  expect(result.data.RegisterUserMutation.error).toBeTruthy();
  expect(result.data.RegisterUserMutation.me).toBeNull();
  expect(sanitizeTestObject(result.data)).toMatchSnapshot();
});

it("should not be register a user with incomplete input", async () => {
  const mutation = `
  mutation RegisterUserMutation($input: UserRegisterInput!) {
    RegisterUserMutation(input: $input) {
      token
      me {
        id
        name
        email
      }
      error
    }
  }
  `;

  const variableValues = {
    input: { email: "test@test.com", password: "123456" },
  };
  const rootValue = {};

  const result = await graphql({
    schema,
    source: mutation,
    rootValue,
    variableValues,
  });

  expect(result.errors).toBeTruthy();
  expect(result.data).toBeUndefined();
  expect(sanitizeTestObject(result.data)).toMatchSnapshot();
});
