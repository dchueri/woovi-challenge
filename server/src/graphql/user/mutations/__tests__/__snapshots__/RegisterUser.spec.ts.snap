// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should not be register a user with incomplete input 1`] = `undefined`;

exports[`should not register a user when email already exists 1`] = `
{
  "RegisterUserMutation": {
    "error": "This email has been registered. Please try again!",
    "me": "EMPTY",
    "token": "EMPTY",
  },
}
`;

exports[`should register a user 1`] = `
{
  "RegisterUserMutation": {
    "error": "EMPTY",
    "me": {
      "email": "test@test.com",
      "id": "FROZEN-ID",
      "name": "test",
    },
    "token": "eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpYXQiOjE2NjQ2MzgwNTl9.l0rCcqhFFsW2ibZPSlb_F2ihCPiTdxrXwOCtdvuyRX4",
  },
}
`;
